### PR TITLE
add context for connect and close functions

### PR DESCRIPTION
**Pull request by bpross**
This adds functions:

- `ConnectCtx`
- `CloseCtx`
- `OnConnectCtx`
- `OnCloseCtx`

that mimic the same functions without `Ctx` but allow passing in a context object. This can be useful for adding telemetry, etc.

I am not a huge fan of the names of things, but since I wanted to maintain backwards compatibility with the library I couldn't really figure out a better way to do this.

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -107,6 +108,10 @@ func New(addr string, spec *iso8583.MessageSpec, mlReader MessageLengthReader, m
 		}
 	}
 
+	if err := opts.Validate(); err != nil {
+		return nil, fmt.Errorf("validating options: %w", err)
+	}
```
Comment (adamdecaf) [inline connection.go]: This doesn't seem related to adding `*Ctx()` methods. What is the need here? Will it break existing behavior?

File: connection.go [hunk 2]
```diff
@@ -188,6 +193,49 @@ func (c *Connection) Connect() error {
...
 }
 
+// ConnectCtx establishes the connection to the server using configured Addr
+func (c *Connection) ConnectCtx(ctx context.Context) error {
+	var conn net.Conn
+	var err error
+
+	if c.conn != nil {
+		return nil
+	}
```
Comment (adamdecaf) [inline connection.go]: Usually Connect would be written in terms of ConnectCtx.

```go
func Connect() error {
   return ConnectCtx(context.Background())
}
func ConnectCtx(ctx context.Context) error {
   // so we only have one implementation of the logic
}
```

https://github.com/golang/go/blob/go1.22.3/src/database/sql/sql.go#L1657-L1676

File: connection.go (hunk 1 above)
Comment (bpross) [reply connection.go]: the idea here is to verify that either the Ctx methods or the original methods are configured but not both. But with rewriting the methods to only have one implementation should take care of this, so I'll remove it.

File: connection.go [hunk 3]
```diff
@@ -170,7 +176,16 @@ func (c *Connection) Connect() error {
 
 	c.run()
 
-	if c.Opts.OnConnect != nil {
+	if c.Opts.OnConnectCtx != nil {
```
Comment (alovak) [inline connection.go]: @bpross I think about two options here.

1. Get rid of `Options/OnConnect` and for backward compatibility, we can update the `OnConnect` function:
```go
func OnConnect(h func(c *Connection) error) Option {
	return func(opts *Options) error {
		opts.OnConnectCtx = func(_ context.Context, c *Connection) error {
                  return h(c)
                }
		return nil
	}
}
```

this is not 100% backward compatible, as if someone directly sets `Options/OnConnect` then the code will be broken. While no one should directly work with this option, it's exportable and we have no control over it now.

2. We can change the code to something like:
```go
	onConnect := c.Opts.OnConnectCtx
	if onConnect == nil && c.Opts.OnConnect != nil {
		onConnect = func(_ context.Context, c *Connection) error {
			return c.Opts.OnConnect(c)
		}
	}

	if onConnect != nil {
		if err := onConnect(ctx, c); err != nil {
			...
```

Comment (bpross) [reply connection.go]: I think the second of your options is preferable and cleans up the code a bit. It reduces branch complexity and keeps it backwards compatible if someone is directly setting `OnConnect`

**Review by alovak — APPROVED** [review]

**Review by adamdecaf — APPROVED** [review]
Should we mention that the Ctx methods take priority if both are non-nil?

**Review by alovak — CHANGES_REQUESTED** [review]
I'm thinking about the context we pass here, and it seems that we don't use it in any of the package functions/methods. The main purpose is to pass it to the user's handlers. But then, I'm like: do we actually need this?

**Comment by adamdecaf** [comment]
Do you want to wire in otel to these contexts? 

https://github.com/moov-io/base/tree/master/telemetry

**Comment by alovak** [comment]
In DB driver, when we pass the context, driver can close the connection if context is Canceled or Done, but here we don't use context and that's my concern. We accept it and pass it through to handlers so they can do something with it. Is it good enough reason for this change?

**Comment by bpross** [comment]
As @adamdecaf posted the whole point of passing through these contexts into the user defined handlers allows the user to pass through any values that they have attached to the context values. In Moov's internal case we attach our telemetry to contexts, so this allows us to attach child spans in the handlers to the root spans from calling `ConnectCtx` or `CloseCtx`. This is useful during startup/shutdown, because we can follow these units of work in telemetry instead of having them dangling. 

Initially I had something like this (and something similar for OnClose):

```
connectionOpts.OnConnect(func(c *connection.Connection) error {
    ctx, span := StartSpan(context.Background())
    defer span.End()
    
    return signOnfunc(ctx, c)
})
```

this gets us most of the way, but produces a "dangling" span, so we cannot associate it with app startup, or if we just tried to reconnect, etc. 

I agree that we can enhance the use of the contexts that are being passed into these methods and look for cancelation, etc. However, given the context of the work that I am doing at the moment, that would fall out of scope for that.

**Review by alovak — APPROVED** [review]

**Review by alovak — APPROVED** [review]

**Comment by adamdecaf** [comment]
Sounds good.

Can you pass the app-startup span we create to these calls? That span's End() is called when NewEnvironment returns.

**Review by adamdecaf — APPROVED** [review]
